joomla/joomla-cms
# [4.0] Filter Parameters in Options

This PR adds the ability to filter parameters in Joomla! configuration options

![Filter Options](https://i.imgur.com/4Y632b2.png)

![Filter Options](https://i.imgur.com/nbIYNTq.gif)


### Summary of Changes
The following files were changed:
**administrator\components\com_config\tmpl\component\default.php** - linked layout file
**administrator\components\com_config\tmpl\application\default.php** - linked layout file
**layouts\joomla\searchtools\filter_options.php** - layout file displaying filter input field
**administrator\language\en-GB\en-GB.ini** - added one language string
**build\media_src\com_config\css\filter-options.css** - added CSS to display search results in Global Configuration in a better way (Global Configuration is problematic part because it includes columns)
**build\media_src\com_config\js\config-filter-options.es6.js** - main Javascript file (thanks to David Jardin and Benjamin Trenkle for overwriting it from jQuey to Vanilla JS).


### Testing Instructions
Because the Javascript file and CSS file should be built by **npm**, this feature must be tested the following way:
https://docs.joomla.org/J4.x:Setting_Up_Your_Local_Environment

After building Javascript and CSS files, just go to Joomla! Global Configuration or Options of some component and type into the filter input field.

### Expected result
You should be able to filter parameters in Joomla! options, see videos:

See YouTube video (Joomla! 4):
[Filter Parameters in Options - Joomla! 4](https://youtu.be/sn3bFggDdzI)

You can test this feature in Joomla! 3.9 with help of the following system plugin:
https://www.phoca.cz/phoca-filter-options-system-plugin

See YouTube video (Joomla! 3.9 - there is no PR in Joomla! 3.9 but system plugin)
[Phoca Filter Options - Joomla! 3.9](https://youtu.be/zKFdgqgbs1M)

### Possible Problems
- For now, I don't know if file names and the file location match Joomla! convention rules, if not, this needs to be changed
- Global Configuration is a problematic part, because of columns. This feature is based on manipulating design parts. Global configuration needs extra CSS and maybe it needs to be improved
- A11y not tested yet
- Possible improvement could be unification of HTML output (Global Configuration / Component Options => #configContent vs. #config-document) and changing HTML Output of Options so there will be less Javascript manipulation.
